### ISSUE TITLE
Panic on invalid URL characters

### ISSUE THREAD

**Issue by cetra3** [actix/actix-web#130]
If you submit a request to an actix web server which doesn't include valid Uri characters, you will get a panic

Steps to replicate:

* Start up the `hello-world` example
* Submit a request via curl with an invalid `^` character:

```
curl 'http://localhost:8080/^'
```

* You will see the following:

```
Starting http server: 127.0.0.1:8080
thread 'arbiter:"b5cc4942-8c9d-4914-a620-15c02e8d0581":"actor"' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidUri(InvalidUriChar)', libcore/result.rs:945:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:207
   3: std::panicking::default_hook
             at libstd/panicking.rs:223
   4: std::panicking::begin_panic
             at libstd/panicking.rs:402
   5: std::panicking::try::do_call
             at libstd/panicking.rs:349
   6: std::panicking::try::do_call
             at libstd/panicking.rs:325
   7: core::ptr::drop_in_place
             at libcore/panicking.rs:72
   8: core::result::unwrap_failed
             at /Users/travis/build/rust-lang/rust/src/libcore/macros.rs:26
   9: <core::result::Result<T, E>>::unwrap
             at /Users/travis/build/rust-lang/rust/src/libcore/result.rs:782
  10: actix_web::server::h1::Reader::parse_message
             at /Users/cetra/git/actix-web/src/server/h1.rs:507
  11: actix_web::server::h1::Reader::parse
             at /Users/cetra/git/actix-web/src/server/h1.rs:460
  12: <actix_web::server::h1::Http1<T, H>>::poll_io
             at /Users/cetra/git/actix-web/src/server/h1.rs:137
  13: <actix_web::server::h1::Http1<T, H>>::poll
             at /Users/cetra/git/actix-web/src/server/h1.rs:119
  14: <actix_web::server::channel::HttpChannel<T, H> as futures::future::Future>::poll
             at /Users/cetra/git/actix-web/src/server/channel.rs:88
  15: <actix_web::server::channel::HttpChannel<T, H> as futures::future::Future>::poll
             at /Users/cetra/git/actix-web/src/server/channel.rs:139
  16: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/future/mod.rs:113
  17: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/task_impl/mod.rs:289
  18: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/task_impl/mod.rs:363
  19: futures::task_impl::std::set
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/task_impl/std/mod.rs:78
  20: <futures::task_impl::Spawn<T>>::enter
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/task_impl/mod.rs:363
  21: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.19/src/task_impl/mod.rs:289
  22: <tokio::executor::current_thread::scheduler::Scheduled<'a, U>>::tick
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/scheduler.rs:343
  23: <tokio::executor::current_thread::scheduler::Scheduler<U>>::tick::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/scheduler.rs:323
  24: <tokio::executor::current_thread::Borrow<'a, U>>::enter::{{closure}}::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:640
  25: tokio::executor::current_thread::CurrentRunner::set_spawn
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:672
  26: <tokio::executor::current_thread::Borrow<'a, U>>::enter::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:639
  27: <std::thread::local::LocalKey<T>>::try_with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:290
  28: <std::thread::local::LocalKey<T>>::with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:244
  29: <tokio::executor::current_thread::Borrow<'a, U>>::enter
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:638
  30: <tokio::executor::current_thread::scheduler::Scheduler<U>>::tick
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/scheduler.rs:323
  31: <tokio::executor::current_thread::Entered<'a, P>>::tick
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:531
  32: <tokio::executor::current_thread::Entered<'a, P>>::turn
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.3/src/executor/current_thread/mod.rs:473
  33: tokio_core::reactor::Core::poll::{{closure}}::{{closure}}::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.15/src/reactor/mod.rs:326
  34: <scoped_tls::ScopedKey<T>>::set
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.1/src/lib.rs:155
  35: tokio_core::reactor::Core::poll::{{closure}}::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.15/src/reactor/mod.rs:325
  36: tokio_executor::global::with_default::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.0/src/global.rs:147
  37: <std::thread::local::LocalKey<T>>::try_with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:290
  38: <std::thread::local::LocalKey<T>>::with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:244
  39: tokio_executor::global::with_default
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.0/src/global.rs:121
  40: tokio_core::reactor::Core::poll::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.15/src/reactor/mod.rs:289
  41: tokio_reactor::with_default::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.0/src/lib.rs:193
  42: <std::thread::local::LocalKey<T>>::try_with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:290
  43: <std::thread::local::LocalKey<T>>::with
             at /Users/travis/build/rust-lang/rust/src/libstd/thread/local.rs:244
  44: tokio_reactor::with_default
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.0/src/lib.rs:185
  45: futures::task_impl::Notify::clone_id
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.15/src/reactor/mod.rs:288
  46: tokio_core::reactor::Core::run
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.15/src/reactor/mod.rs:264
  47: actix::arbiter::Arbiter::new::{{closure}}
             at /Users/cetra/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.5.5/src/arbiter.rs:89
```

### What I'd expect to happen

My suggestion would be that either the invalid characters are passed through and allowed, OR they are not allowed and a`400` is returned to the client.  

If this could be configurable, it would be awesome, as some legacy clients I deal with can't be changed.

**Comment by fafhrd91** [comment]
thanks for report!

i dont think it is possible to bypass `http::Uri` parsing errors, but i'll check. at least it should be 400.

**Comment by cetra3** [comment]
Yeah I think that there are some security ramifications around allowing invalid chars through anyway.  It's a nice to have if we can ignore it, but I'm not too hung up on it.

**Comment by fafhrd91** [comment]
should be fixed by 5a25fd9
